hypervisor/src/cpu.rs

```diff
@@ -295,3 +294,15 @@ pub trait Vcpu: Send + Sync {
     ///
     fn set_cpu_state(&self, state: &CpuState) -> Result<()>;
 }
+
+pub trait VcpuRun {
+    fn id(&self) -> u8;
+    fn mmio_read(&self, addr: u64, data: &mut [u8]) -> Result<bool>;
+    fn mmio_write(&self, addr: u64, data: &[u8]) -> Result<bool>;
+    #[cfg(target_arch = "x86_64")]
+    fn pio_in(&self, addr: u64, data: &mut [u8]) -> Result<bool>;
+    #[cfg(target_arch = "x86_64")]
+    fn pio_out(&self, addr: u64, data: &[u8]) -> Result<bool>;
+    #[cfg(target_arch = "x86_64")]
+    fn eoi(&self, vector: u8) -> Result<bool>;
+}
```

Comment:
Thanks for the proposal @liuw .

What we're trying to do here is abstracting the hypervisor specific VM exit handling into a set of basic operations through this trait. Overall, the end goal is to have an hypervisor agnostic vmm/src/cpu.rs `run()` implementation. Is that a fair description?

I definitely like your proposal, but I think the API cut would be simpler if we could avoid having to implement an hypervisor defined trait from the VMM. IMHO hypervisor traits should be implemented from hypervisor implementations, not from the VMM itself. Also, this trait makes assumption about how the VMM handle its device model through specific IO commands.

I'm proposing a slight variation of this approach by passing hypervisor agnostic vm exits results from the hypervisor trait back to the vmm. See:
https://github.com/sameo/cloud-hypervisor-1/commit/7affeade54052afb5e414d6cc86f14d1c30ce74a

for an implementation example.

The VMM `vcpu.run()` loop is KVM and hypervisor agnostic with that variation, but the VMM does not have to implement an hypervisor defined trait. In particular, it does not have to make that call:

```
self.fd.run(self).map_err(|e| Error::VcpuRun(e.into()))
```

where we end up having to pass ourselves as an inner argument.

We're basically splitting the VM exit handling into hypervisor specific bits vs VMM ones, and we let each VMM decide how IO related exits should be handled, emulated or ignored.

 What do you think?

Reply:
> 
> 
> Thanks for the proposal @liuw .
> 
> What we're trying to do here is abstracting the hypervisor specific VM exit handling into a set of basic operations through this trait. Overall, the end goal is to have an hypervisor agnostic vmm/src/cpu.rs `run()` implementation. Is that a fair description?
> 

Yes. This is a fair description.

> I definitely like your proposal, but I think the API cut would be simpler if we could avoid having to implement an hypervisor defined trait from the VMM. IMHO hypervisor traits should be implemented from hypervisor implementations, not from the VMM itself. Also, this trait makes assumption about how the VMM handle its device model through specific IO commands.
> 

Right. I get your concerns here.

> I'm proposing a slight variation of this approach by passing hypervisor agnostic vm exits results from the hypervisor trait back to the vmm. See:
> [sameo@b32dbe6](https://github.com/sameo/cloud-hypervisor-1/commit/b32dbe6ec353911e5159330752dab0803892c051)
> 
> for an implementation example.
> 
> The VMM `vcpu.run()` loop is KVM and hypervisor agnostic with that variation, but the VMM does not have to implement an hypervisor defined trait. In particular, it does not have to make that call:
> 
> ```
> self.fd.run(self).map_err(|e| Error::VcpuRun(e.into()))
> ```
> 
> where we end up having to pass ourselves as an inner argument.
> 
> We're basically splitting the VM exit handling into hypervisor specific bits vs VMM ones, and we let each VMM decide how IO related exits should be handled, emulated or ignored.
> 
> What do you think?

Your approach works for me too. Thanks for the detailed example code. I wasn't sure whether I got your VcpuRun trait idea correctly so this patch was mostly aiming to start this discussion.

Now that we're on the same page. We (MSFT) will produce some more patches later based on this design.

Wei.

Reply:
> > What do you think?
> 
> Your approach works for me too. Thanks for the detailed example code. I wasn't sure whether I got your VcpuRun trait idea correctly so this patch was mostly aiming to start this discussion.
> 

Yes, and it did. I think you got my `VcpuRun` trait idea correctly, and it's when I saw your actual implementation that I thought about how it could be improved. Thanks for starting that effort.

> Now that we're on the same page, we (MSFT) will produce some more patches later based on this design.
> 

Looking forward to getting those. Thanks.

Reply:
Alright. I will submit a PR shortly to get the VmExit abstraction merged first. Thanks.